scripts/ai/AIDriveStrategyAttachHeader.lua

```diff
@@ -0,0 +1,306 @@
+
+AIDriveStrategyAttachHeader.myStates = {
+    WAITING_FOR_DETACH_TO_FINISH = {},
+    DRIVING_AWAY_FROM_HEADER = {}, --- Drives a small distance forwards after detaching
+    DRIVING_TO_HEADER = {}, --- Drives close to the header with the pathfinder
+    DRIVING_TO_ATTACH_CUTTER = {}, --- Drives the last few meters to the cutter for attaching.
+    WAITING_FOR_ATTACH_TO_START = {}, --- Waits a bit, to make sure the harvester stands still.
+    WAITING_FOR_ATTACH_TO_FINISH = {}, --- Waits until the attach animation finished.
+    REVERSING_FROM_CUTTER = {} --- If the cutter was on an extern trailer, drive back a bit.
+}
+AIDriveStrategyAttachHeader.MODES = {
+    ATTACH_HEADER_FROM_ATTACHED_TRAILER = 1, --- The cutter/header is on an extern trailer.
+    ATTACH_HEADER_WITH_WHEELS_ATTACHED = 2, --- The cutter/header is foldable with attached wheels for transport without an extra trailer.
+}
+AIDriveStrategyAttachHeader.DRIVING_AWAY_FROM_HEADER_FORWARD_DISTANCE = 6
+
+
+function AIDriveStrategyAttachHeader.new(customMt)
+    if customMt == nil then
+        customMt = AIDriveStrategyAttachHeader_mt
+    end
+    local self = AIDriveStrategyCourse.new(customMt)
+    AIDriveStrategyCourse.initStates(self, AIDriveStrategyAttachHeader.myStates)
+    self.state = self.states.INITIAL
+    self.debugChannel = CpDebug.DBG_FIELDWORK
+    self.mode = self.MODES.ATTACH_HEADER_FROM_ATTACHED_TRAILER
+    return self
+end
+
+--- The fieldwork course is not needed for this strategy.
+function AIDriveStrategyAttachHeader:isGeneratedCourseNeeded()
+    return false
+end
+
+function AIDriveStrategyAttachHeader:delete()
+    if self.cutterNode then 
+        CpUtil.destroyNode(self.cutterNode)
+    end
+    AIDriveStrategyAttachHeader:superClass().delete(self)
+end
+
+function AIDriveStrategyAttachHeader:initializeImplementControllers(vehicle)
+
+    if AIUtil.hasCutterAsTrailerAttached(vehicle) then 
+        self.mode = self.MODES.ATTACH_HEADER_WITH_WHEELS_ATTACHED
+        self:debug("A cutter/header with foldable wheels is attached at the back.")
+        self.trailer = AIUtil.getImplementWithSpecialization(self.vehicle, Cutter)
+    else 
+        self.mode = self.MODES.ATTACH_HEADER_FROM_ATTACHED_TRAILER
+        self:debug("A cutter/header on a separate trailer is attached on the back.")
+        self.trailer = AIUtil.getImplementWithSpecialization(self.vehicle, DynamicMountAttacher)
+        self.dynamicMountAttacherController = DynamicMountAttacherController(vehicle, self.trailer)
+    end
+    self.attachableController = AttachableController(vehicle, self.trailer)
+    self.attacherJointController = AttacherJointController(vehicle, vehicle)
+    self:appendImplementController(self.attacherJointController)
+    local trailerAreaNode = createTransformGroup("tempTrailerNode")
+    self.trailerAreaToAvoid =  PathfinderUtil.NodeArea(trailerAreaNode, -self.trailer.size.width/2, -self.trailer.size.length/2, self.trailer.size.width, self.trailer.size.length)
```

Comment:
I might be wrong, and anyway, may not be a problem but this assumes that the cutter node is at the root node of the trailer and the root node is exactly in the middle of trailer (they sometimes have a lengthOffset for the size)

Reply:
A good point I am not even sure if this check is even needed. I guess I will put a comment there and we will see if it breaks from a user test.

Reply:
Yeah, you right, the pathfinder should detect a collision there and avoid it anyway.